### PR TITLE
West Midlands | 26 March SDC | Iswat Bello | Sprint 2 | chat app

### PR DESCRIPTION
**Pull request by Iswanna**
## Learners, PR Template

Self checklist

- [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
- [x] My changes meet the requirements of the task
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)

## Changelist
This PR implements a fully functional real-time chat application with a Node.js/Express backend and vanilla JavaScript frontend. The application supports sending messages, viewing chat history, and liking messages with real-time updates using long polling.

## Deployed sites
- A link to the [deployed frontend long polling](https://iswanna-chat-app-frontend-long-polling.hosting.codeyourfuture.io) on the CYF hosting environment.
- A link to the [deployed backend](https://iswanna-chat-app-backend.hosting.codeyourfuture.io/messages) on the CYF hosting environment.

## Additional Feature implemented
- **Interactive Reaction System**: I implemented a "Like" button for each message. This required creating a custom API endpoint (/messages/:id/like) and logic to update the state of existing messages without re-rendering the entire chat.

- **Performance Optimisation (Long Polling)**: To make the chat feel "live" without overworking the server, I implemented Long Polling. This ensures messages appear instantly while reducing the number of empty requests.

### CODE REVIEW

**Comment by cjyuan** [comment]
- The content in the "Changes Made" section is better kept in a `.md` file.

- Can you include these in the PR description (as requested in the spec)?
    - A link to the deployed frontend on the CYF hosting environment
    - A link to the deployed backend on the CYF hosting environment

- Can you also list the extra features you implemented? (The spec says the app must also support at least one additional feature)

File: chat-app/backend/server.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// Middleware to parse JSON request body
+app.use(express.json());
+
+app.post("/messages", (req, res) => {
+  const { text, sender } = req.body;
+
+  // The validation
+  if (!text || text.length === 0 || !sender || sender.length === 0) {
+    return res.status(400).send("Please provide both text and a sender name.");
+  }
```
Comment (cjyuan) [inline chat-app/backend/server.js:82]: It's good practice to validate all request inputs.

Note:  A more robust implementation 
- Should also handle the case where `req.body` is `null`, which would cause `const { text, sender } = req.body;` to throw.
- Should not assume `text` and `sender` are strings (if they existed)

In this exercise, let's assume the request is always coming from the code in `script.js`. **So no change needed.**

File: chat-app/frontend/script.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+const formElement = document.getElementById("chat-form");
+const senderElement = document.getElementById("chat-sender");
+const messageElement = document.getElementById("chat-message");
+
+formElement.addEventListener("submit", async (event) => {
+  event.preventDefault();
+
+  const senderValue = senderElement.value;
+  const messageValue = messageElement.value;
```
Comment (cjyuan) [inline chat-app/frontend/script.js]: No input validation?

File: chat-app/frontend/script.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  try {
+    // Send the data
+    const response = await fetch(
+      "https://iswanna-chat-app-backend.hosting.codeyourfuture.io/messages",
```
Comment (cjyuan) [inline chat-app/frontend/script.js]: To make the app more configurable, consider keeping the domain name in a global named constant.

File: chat-app/backend/server.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  while (callBacksForNewMessages.length > 0) {
+    // take the last function out the array
+    const callback = callBacksForNewMessages.pop();
+
+    // run that function using the newMessage as argument
+    callback([newMessage]);
+  }
+
+  // Finally, respond to the person who actually sent the POST request
+  res.status(201).send(newMessage);
```
Comment (cjyuan) [inline chat-app/backend/server.js:116]: Why embed the received message in the response? It is not being used in `script.js`.

File: chat-app/backend/server.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    return res.status(400).send("Please provide both text and a sender name.");
+  }
+
+  // Create the message object
+  const newMessage = {
+    id: messages.length,
+    sender: sender,
+    text: text,
+    likes: 0,
+    dislikes: 0,
```
Comment (cjyuan) [inline chat-app/backend/server.js:90]: What's the `dislikes` property for? It's not used anywhere in the app.

File: chat-app/backend/server.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  if (sinceValue !== undefined) {
+    sinceId = Number(sinceValue);
+  } else {
+    sinceId = -1;
+  }
+
+  const messagesSinceId = messages.filter((message) => message.id > sinceId);
+
+  if (messagesSinceId.length === 0) {
+    callBacksForNewMessages.push((value) => res.send(value));
```
Comment (cjyuan) [inline chat-app/backend/server.js:134]: What do you expect to happen if no new message arrives within 30 minutes?

File: chat-app/backend/server.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+  messageWithIdAsNumber.likes += 1;
+
+  if (messageWithIdAsNumber) {
+    while (callBacksForNewMessages.length > 0) {
+      const callback = callBacksForNewMessages.pop();
+
+      callback([messageWithIdAsNumber]);
+    }
+    res.status(200).send(messageWithIdAsNumber);
```
Comment (cjyuan) [inline chat-app/backend/server.js]: Some of the message properties are not needed to update `likes` count. Can you improve this code to send only the necessary data?

File: chat-app/backend/server.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+app.post("/messages/:id/like", (req, res) => {
+  // Get the id from the URL
+  const idFromUrl = req.params.id;
+
+  //convert to number
+  const idAsNumber = Number(idFromUrl);
+
+  const messageWithIdAsNumber = messages.find(
+    (message) => message.id === idAsNumber,
+  );
```
Comment (cjyuan) [inline chat-app/backend/server.js]: Is there a more efficient way to locate the desired message object by its id?

**Comment by Iswanna** [comment]
> * The content in the "Changes Made" section is better kept in a `.md` file.
> * Can you include these in the PR description (as requested in the spec)?
>   
>   * A link to the deployed frontend on the CYF hosting environment
>   * A link to the deployed backend on the CYF hosting environment
> * Can you also list the extra features you implemented? (The spec says the app must also support at least one additional feature)

I have updated the PR description based on your feedback. Thank you.

File: chat-app/backend/server.js (hunk 1 above)
Comment (Iswanna) [reply chat-app/backend/server.js:82]: Thank you for the feedback.

I have updated the backend validation to be more robust as suggested.
While the frontend script.js currently sends valid data, I used Thunder Client in VS Code to test the API directly with edge cases. This helped me identify two main risks:

**Server Crashes:** If the request body is missing or null, the server returns a 500 error when attempting to destructure the input.
**Data Distortion:** I found that without type-checking, the server would accept non-string data (like arrays or numbers). This would distort the data being stored in the backend and could cause the frontend to display messages incorrectly.

File: chat-app/frontend/script.js (hunk 2 above)
Comment (Iswanna) [reply chat-app/frontend/script.js]: I have updated the code by adding frontend validation to the submit listener. It now trims the inputs and prevents the fetch request if the name or message is empty, providing an immediate alert to the user.

File: chat-app/frontend/script.js (hunk 3 above)
Comment (Iswanna) [reply chat-app/frontend/script.js]: I have implemented your feedback. Thank you.

File: chat-app/backend/server.js (hunk 4 above)
Comment (Iswanna) [reply chat-app/backend/server.js:116]: Thank you for the feedback.

I embedded the received message in the response so that the client can verify the integrity of the transaction. It provides a clear confirmation that the data was received and stored exactly as intended. 

What do you think?

File: chat-app/backend/server.js (hunk 5 above)
Comment (Iswanna) [reply chat-app/backend/server.js:90]: Thank you for catching that. The dislikes property was part of what I intended to implement, but since I haven't implemented it yet, I will remove it to keep the data model clean and efficient.

File: chat-app/backend/server.js (hunk 4 above)
Comment (cjyuan) [reply chat-app/backend/server.js:116]: The rationale would probably make more sense when the client is a third party. If both the client and server are implemented by the same party, then even simply returning "OK" (or any similar indication) would be as good a confirmation.

There are other reasons why in some application developers practice returning the post-processed message in the response (for a POST request to add new message). Feel free to use AI to explore the reasons behind this design choice so you can make better decisions about what to return in future work.

No change needed.

File: chat-app/backend/server.js (hunk 6 above)
Comment (Iswanna) [reply chat-app/backend/server.js:134]: I expect the connection to be closed by the hosting environment or the browser. 
To handle this, I added **Unique Client IDs**:

**Frontend**: If the connection times out, my code catches the error and retries using the same ID.
**Backend**: I replaced the array with an Object. The server uses the Client ID to overwrite the old connection with the new one.

This ensures the server's memory doesn't fill up with 'ghost' connections, even if no messages arrive for 30 minutes.

File: chat-app/backend/server.js (hunk 7 above)
Comment (Iswanna) [reply chat-app/backend/server.js]: I have updated the code to send a subset object containing only the `id` and `likes` properties. This makes the API response more efficient by sending only the data needed for the UI update, rather than the entire message. 

Thank you for the feedback.

File: chat-app/backend/server.js (hunk 8 above)
Comment (Iswanna) [reply chat-app/backend/server.js]: Yes, there is. I replaced the .find() method with direct bracket access `(messages[idAsNumber])`.

Thank you for the feedback.

File: chat-app/backend/server.js (hunk 4 above)
Comment (Iswanna) [reply chat-app/backend/server.js:116]: Thank you so much for the detailed explanation. I will also do more research as you have suggested.

File: chat-app/backend/server.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    if (!clientId) {
+      return res.send([]);
+    }
+    callBacksForNewMessages[clientId] = (value) => {
+      try {
+        res.send(value);
+      } catch (e) {
+        delete callBacksForNewMessages[clientId];
+      }
+    };
```
Comment (cjyuan) [inline chat-app/backend/server.js]: Note: The new logic involving `clientId` might complicate things.

If the client becomes unreachable for any reason, `res.send(value)` may not throw an error. As a result, the server may retain references to disconnected clients and gradually leak memory.

Your previous approach (which you clear the waiting room after sending a response) is actually quite robust. If a client is unreachable when the server calls `res.send(value)`, the response simply never reaches the client.  As long as the client can detect lost connection, it can always issue a new request to re-establish communication.

Note: You can keep this mechanism as is. No change needed.

File: chat-app/frontend-polling/script.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+let lastIdSeen = -1;
+
+async function getAllMessages() {
+  try {
+    const response = await fetch(
+      `${API_BASE_URL}/messages?since=${lastIdSeen}&clientId=${myClientId}`,
+    );
+
+    const data = await response.json();
```
Comment (cjyuan) [inline chat-app/frontend-polling/script.js:20]: A better and more robust pattern is:
```
try {
  const res = await fetch(url, { signal });

  if (!res.ok) {
    throw new Error(`HTTP ${res.status}`);
  }

  const data = await res.json();

} catch (err) {
  console.error(err);
}
```

This way, the error caught in the `catch` block would be an error about the HTTP response and not a decoding error thrown by `res.json()`.

File: chat-app/frontend/script.js [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+    const data = await response.json();
+
+    const messageContainer = document.getElementById("all-messages");
+
+    data.forEach((message) => {
+      const elementId = "msg-" + message.id;
+
+      const existingElement = document.getElementById(elementId);
+
```
Comment (cjyuan) [inline chat-app/frontend/script.js]: If the presentation logic (lines 15 to 61) is kept in a separate function that take messages as its parameter, it would make the rendering logic easier to develop and test independently.

File: chat-app/frontend/script.js [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+        messageContainer.appendChild(newElement);
+
+        lastIdSeen = message.id;
+      }
+    });
+    setTimeout(getAllMessages, 0);
+  } catch (error) {
+    setTimeout(getAllMessages, 0);
+    console.error("Error fetching messages:", error);
+  }
```
Comment (cjyuan) [inline chat-app/frontend/script.js]: Could consider using a `finally` block:

```
try {
   ...
} catch (error) {
}
finally {
  setTimeout(getAllMessages, 0);
}
```

**Review by cjyuan — COMMENTED** [review]
I just have a few more suggestions.

File: chat-app/backend/server.js (hunk 9 above)
Comment (Iswanna) [reply chat-app/backend/server.js]: I have decided to revert to the array-based approach to keep my code robust. Thank you for the feedback.

File: chat-app/frontend-polling/script.js (hunk 10 above)
Comment (Iswanna) [reply chat-app/frontend-polling/script.js:20]: I have implemented the robust fetch pattern in `getAllMessages` as suggested. It now checks `!response.ok` and throws a specific `HTTP` error, which ensures that the catch block handles server errors correctly and avoids decoding errors in `response.json()`. 

Thanks for the detailed feedback

File: chat-app/frontend/script.js (hunk 11 above)
Comment (Iswanna) [reply chat-app/frontend/script.js]: I have refactored the code to separate the presentation logic into its own function, `renderMessages(data)`, as suggested.

File: chat-app/frontend/script.js (hunk 12 above)
Comment (Iswanna) [reply chat-app/frontend/script.js]: I have refactored the `getAllMessages` function to use a finally block for the recursive `setTimeout` call.

Thank you for the feedback.

File: chat-app/frontend/script.js [hunk 13]
```diff
@@ -1,69 +1,79 @@
...
+    renderMessages(data);
+  } catch (error) {
+    console.error("Error fetching messages:", error);
+  } finally {
+    setTimeout(getAllMessages, 0);
+  }
+}
 
-      const existingElement = document.getElementById(elementId);
+function renderMessages(data) {
```
Comment (cjyuan) [inline chat-app/frontend/script.js]: `messages` would probably a clearer name for this parameter.

**Review by cjyuan — COMMENTED** [review]
Changes look good. Well done.

File: chat-app/frontend/script.js (hunk 13 above)
Comment (Iswanna) [reply chat-app/frontend/script.js]: I have replaced the parameter `data` with `messages` in the `renderMessages` function.

**Comment by Iswanna** [comment]
> Changes look good. Well done.

Thank you so much for the detailed PR review. I really appreciate it.

**Comment by illicitonion** [comment]
Closing PR because the SDC run has finished. Feel free to re-open if you're still working on it.
